applications/tari_base_node/src/builder.rs

```diff
@@ -495,43 +493,155 @@ where
             };
 
             if bootstrapped {
+                debug!(
+                    target: LOG_TARGET,
+                    "Initial sync achieved and starting with transaction and UTXO validation protocols.",
+                );
                 let _ = transaction_service_handle
                     .restart_broadcast_protocols()
                     .await
                     .map_err(|e| {
                         error!(
                             target: LOG_TARGET,
-                            "Problem restarting broadcast protocols in the Transaction Service"
+                            "Problem restarting broadcast protocols in the Transaction Service: {}", e
                         );
                         e
                     });
-
                 let _ = transaction_service_handle
                     .restart_transaction_protocols()
                     .await
                     .map_err(|e| {
                         error!(
                             target: LOG_TARGET,
-                            "Problem restarting transaction negotiation protocols in the Transaction Service"
+                            "Problem restarting transaction negotiation protocols in the Transaction Service: {}", e
                         );
                         e
                     });
+
+                #[derive(PartialEq, Debug)]
+                enum ValidationTask {
+                    InvalidOutputs,
+                    UnspentOutputs,
+                }
+                let mut event_stream = oms_handle.get_event_stream_fused();
+                let mut validation = ValidationTask::InvalidOutputs;
+                let mut end_protocol = false;
+                let mut delay = delay_for(config_base_node_query_timeout_clone * 2).fuse();
+                loop {
+                    match validation {
+                        ValidationTask::InvalidOutputs => {
```

Comment:
Revalidating the invalid UTXO's is not a critical task. This is why it doesn't have its own explicit API just yet. I don't think you need to explicitly wait for it to complete, this is partly why this code is so complex. If you want to do it explicitly then it should be removed from the `set_base_node_public_key` method and made explicit so you can monitor the event bus for it but I do not think this is needed. Don't wait for it to complete after setting the public key.

Reply:
I do not agree that only the unspent UTXOs validation should be monitored; both tasks are important and can be perfectly well monitored as the code stands. If removing revalidating invalid UTXOs from the `set_base_node_public_key` method is important, that could be another PR.

Reply:
This code is clunky because of this issue. If you really feel that it is important enough to be monitored here then you should make the change and use the protocol IDs to properly monitor these instances of the validation protocols. It is part of the reason why you have had to introduce an extraneous delay to this code that is absolutely not required.

Reply:
I can change the logic, but it will not any more efficient. The safety net delay for the `futures::select! {` loop only comes into play if the validation protocol stops sending events altogether.

Reply:
Fixed